File: benchmarks/vqe/qcor/qcor_trotter.py
from qcor import *
import time
@qjit
def trotter_circ(q : qreg, exp_args: List[PauliOperator], n_steps: int):
  for i in range(n_steps):
    for i, exp_arg in enumerate(exp_args):
      exp_i_theta(q, 1.0, exp_args[i])

def heisenberg_ham(n_qubits):
  Jz = 1.0
  h = 1.0
  H = 0.0 * X(0)
  for i in range(n_qubits):
    H = H - h * X(i)
  for i in range(n_qubits - 1):
    H = H - Jz * (Z(i) * Z(i + 1))
  return H

n_qubits = [10, 20, 50, 100]
nbSteps = 100

for nbQubits in n_qubits:
  ham_op = heisenberg_ham(nbQubits)
  op_terms = ham_op.getNonIdentitySubTerms()
  q = qalloc(nbQubits)
  start = time.time()
  comp = trotter_circ.extract_composite(q, op_terms, nbSteps)
  end = time.time()
  print("n instructions =", comp.nInstructions(), "; Kernel eval time:", end - start, " [secs]")

File: benchmarks/vqe/qcor/qcor_trotter.cpp
#include <chrono>

// Trotter evolution
__qpu__ void trotter_evolve(qreg q,
                            std::vector<std::shared_ptr<Observable>> &exp_args,
                            int n_steps) {
  for (int i = 0; i < n_steps; ++i) {
    for (auto &exp_arg : exp_args) {
      exp_i_theta(q, 1.0, exp_arg);
    }
  }
}

int main() {
  // Helper to create general Heisenberg Hamiltonian
  const auto heisenberg_ham = [](int nbQubits) {
    const double Jz = 1.0;
    const double h = 1.0;
    auto H = -h * X(0);
    for (int i = 1; i < nbQubits; ++i) {
      H = H - h * X(i);
    }
    for (int i = 0; i < nbQubits - 1; ++i) {
      H = H - Jz * (Z(i) * Z(i + 1));
    }
    return H;
  };

  const std::vector<int> n_qubits{5, 10, 20, 50, 100};
  const int nbSteps = 100;

  for (const auto &nbQubits : n_qubits) {
    auto ham_op = heisenberg_ham(nbQubits);
    auto op_terms = ham_op.getNonIdentitySubTerms();
    auto q = qalloc(nbQubits);
    auto start = std::chrono::high_resolution_clock::now();
    auto nInsts = trotter_evolve::n_instructions(q, op_terms, nbSteps);
    auto stop = std::chrono::high_resolution_clock::now();
    auto duration =
        std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
    const double elapsed_secs = (double)duration.count() / 1e6;
    print("n_qubits =", nbQubits, "; n instructions =", nInsts,
          "; Kernel eval time:", elapsed_secs, " [secs]");
  }
}
